File: api-quarkus/src/main/java/br/ufrn/musi/adaptadores/busca/Dtos.java
package br.ufrn.musi.adaptadores.busca;

import br.ufrn.musi.dominio.Dominio.*;
import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.List;

/**
 * DTOs da fronteira. Espelham contratos/filtro.schema.json.
 */
public final class Dtos {
    private Dtos() {}

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record FiltroDto(String tipo, String dimensao, String valor,
                            List<FiltroDto> opcoes, List<FiltroDto> exigencias,
                            FiltroDto filtro, Integer ano) {

        public static FiltroDto tem(String d, String v) {
            return new FiltroDto("tem", d, v, null, null, null, null);
        }
    }

    public record FacetaDto(String dimensao, String valor) {}

    public record ObraDto(String id, String titulo, String artista, int ano,
                          List<FacetaDto> facetas) {}

    public static Obra paraDominio(ObraDto d) {
        return new Obra(d.id(), d.titulo(), d.artista(), d.ano(),
            d.facetas().stream().map(f -> new Faceta(f.dimensao(), f.valor())).toList());
    }

    public static ObraDto paraDto(Obra o) {
        return new ObraDto(o.id(), o.titulo(), o.artista(), o.ano(),
            o.facetas().stream().map(f -> new FacetaDto(f.dimensao(), f.valor())).toList());
    }

    /**
     * Domínio → DTO.
     *
     * Switch exaustivo sobre a interface selada: acrescentar um construtor de Filtro
     * faz este método parar de compilar. É o equivalente exato do `when` do lado
     * Kotlin, em `Dtos.kt`.
     */
    public static FiltroDto paraDto(Filtro f) {
        return switch (f) {
            case Tem t    -> new FiltroDto("tem", t.dimensao(), t.valor(), null, null, null, null);
            case Ou o     -> new FiltroDto("ou", null, null,
                                 o.opcoes().stream().map(Dtos::paraDto).toList(), null, null, null);
            case E e      -> new FiltroDto("e", null, null, null,
                                 e.exigencias().stream().map(Dtos::paraDto).toList(), null, null);
            case Exceto x -> new FiltroDto("exceto", null, null, null, null, paraDto(x.filtro()), null);
            case Ate a    -> new FiltroDto("ate", null, null, null, null, null, a.ano());
        };
    }

    /** DTO → domínio, na fronteira HTTP pública. */
    public static Filtro paraDominio(FiltroDto d) {
        return switch (d.tipo()) {
            case "tem"    -> new Tem(d.dimensao(), d.valor());
            case "ou"     -> new Ou(d.opcoes() == null ? List.of()
                                 : d.opcoes().stream().map(Dtos::paraDominio).toList());
            case "e"      -> new E(d.exigencias() == null ? List.of()
                                 : d.exigencias().stream().map(Dtos::paraDominio).toList());
            case "exceto" -> new Exceto(paraDominio(d.filtro()));
            case "ate"    -> new Ate(d.ano());
            default -> throw new IllegalArgumentException("tipo de filtro desconhecido: " + d.tipo());
        };
    }
}

File: api-ktor/src/main/kotlin/br/ufrn/musi/adaptadores/busca/Dtos.kt
package br.ufrn.musi.adaptadores.busca

import br.ufrn.musi.dominio.*
import kotlinx.serialization.Serializable

/**
 * DTOs da fronteira. Espelham contratos/filtro.schema.json.
 *
 * Por que não usar `Filtro` do domínio direto na serialização? Porque isso faria o
 * formato JSON público virar refém da modelagem interna, e vice-versa: renomear um
 * campo do domínio quebraria todos os clientes.
 */
@Serializable
data class FiltroDto(
    val tipo: String,
    val dimensao: String? = null,
    val valor: String? = null,
    val opcoes: List<FiltroDto>? = null,
    val exigencias: List<FiltroDto>? = null,
    val filtro: FiltroDto? = null,
    val ano: Int? = null,
)

@Serializable
data class FacetaDto(val dimensao: String, val valor: String)

@Serializable
data class ObraDto(
    val id: String,
    val titulo: String,
    val artista: String,
    val ano: Int,
    val facetas: List<FacetaDto>,
    val mbid: String? = null,             // ADR-0003: identidade global, ausente até conciliar
    val mbidComposicao: String? = null,
)

fun ObraDto.paraDominio() =
    Obra(id, titulo, artista, ano, facetas.map { Faceta(it.dimensao, it.valor) }, mbid, mbidComposicao)

fun Obra.paraDto() =
    ObraDto(id, titulo, artista, ano, facetas.map { FacetaDto(it.dimensao, it.valor) }, mbid, mbidComposicao)

/**
 * Domínio → DTO.
 *
 * O `when` é exaustivo sobre a interface selada: acrescentar um construtor de Filtro
 * faz esta função parar de compilar, e o compilador aponta onde falta tratar.
 * Ver ADR-0002.
 */
fun Filtro.paraDto(): FiltroDto = when (this) {
    is Filtro.Tem    -> FiltroDto("tem", dimensao = dimensao, valor = valor)
    is Filtro.Ou     -> FiltroDto("ou", opcoes = opcoes.map { it.paraDto() })
    is Filtro.E      -> FiltroDto("e", exigencias = exigencias.map { it.paraDto() })
    is Filtro.Exceto -> FiltroDto("exceto", filtro = filtro.paraDto())
    is Filtro.Ate    -> FiltroDto("ate", ano = ano)
}

fun FiltroDto.paraDominio(): Filtro = when (tipo) {
    "tem"    -> Filtro.Tem(requireNotNull(dimensao), requireNotNull(valor))
    "ou"     -> Filtro.Ou(opcoes.orEmpty().map { it.paraDominio() })
    "e"      -> Filtro.E(exigencias.orEmpty().map { it.paraDominio() })
    "exceto" -> Filtro.Exceto(requireNotNull(filtro).paraDominio())
    "ate"    -> Filtro.Ate(requireNotNull(ano))
    else     -> throw IllegalArgumentException("tipo de filtro desconhecido: $tipo")
}
